### DISCUSSION TITLE
Is it possible to set lower language level without installing the jdk

### DISCUSSION THREAD

**Discussion by tianfudhe** [redhat-developer/vscode-java#2788] [Q&A]
I have only jdk17 installed. 
When I set the project to 1.8(in maven pom.xml), and set "java.configuration.runtimes" with 1.8 having jdk17's path, i.e.,
```
{
  "name": "JavaSE-1.8",
  "path": "/home/tianfu/rsd/jdk17"
}
```
The language server still check the syntax with level17. To address this, I have to install jdk8, and set the runtime path to it.

Is it possible to set level8 with jdk17? So that I only need to install jdk17 to support all projects lower than java17.

**Comment by Eskibear** [comment]
AFAIK, you were on the right way to configure it. @testforstephen @rgrunber Any thoughts? Is it a regression?

**Reply by rgrunber** [reply]
It might be. When we get the project settings from the language server at : 

https://github.com/redhat-developer/vscode-java/blob/6d10e1a00c9ec3628f8934e71be92c1a04acc540/src/runtimeStatusBarProvider.ts#L163

The value comes from https://github.com/eclipse/eclipse.jdt.ls/blob/c9a2d051fc084f2d658c7326956c9698d12bee46/org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/commands/ProjectCommand.java#L133 . I don't see us ever taking into account the execution environment the user has set.

Was `java.configuration.runtimes` ever meant to be used in this way though ? (ie. other than pointing to a JDK of that version)

**Update :** Actually, if you look at https://github.com/eclipse/eclipse.jdt.ls/blob/c9a2d051fc084f2d658c7326956c9698d12bee46/org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/JVMConfigurator.java#L293, I think we get the compliance directly from the JDK installation, and not basing it off the execution environment name provided.

**Comment by Eskibear** [comment]
> Was java.configuration.runtimes ever meant to be used in this way though ? (ie. other than pointing to a JDK of that version)

@rgrunber  It is used to picking JDK you want to use to run your projects. For Maven projects, if you change source level in pom.xml, it eventually modifies jdt compiler settings in `.settings/org.eclipse.jdt.core.prefs`. But the actual JDK corresponding to the source level can be configured by `java.configuration.runtimes`. AFAIK there's no convenient way to explicitly set source level, esp. for projects without build tools. 

> When I set the project to 1.8(in maven pom.xml)

@tianfudhe Sorry I misunderstand the question. Since it's a maven project, I think you don't need any specific setting to get it working. JDK 17 should work well with source level 1.8 as long as you set
```
<maven.compiler.source>1.8</maven.compiler.source>
<maven.compiler.target>1.8</maven.compiler.target>
```
